### PR TITLE
[Android] Add GUI authentication key in settings

### PR DESCRIPTION
**Pull request by talregev**
Add GUI authentication key in settings
Small task inside #1699 #4232

Screenshot_20210301-152744_BOINC
2_Screenshot_20210223-182322_BOINC
Screenshot_20210223-190359_BOINC

### CODE REVIEW

**Comment by davidpanderson** [comment]
What's the use case?  Does it justify the slight decrease in security?

**Comment by AenBleidd** [comment]
@davidpanderson,
> What's the use case?

It's the part of #1699 ticket. It's a feature a lot of people asking for.

> Does it justify the slight decrease in security?

Yes. By default there will be the current behavior where no remote monitoring allowed, so the user will have to turn it on manually. We can add a notification that this decrease the whole security level.

**Comment by talregev** [comment]
password protected field
screenshots:
Screenshot_20210223-130756_BOINC
Screenshot_20210223-130810_BOINC

**Comment by AenBleidd** [comment]
I simplified a task a bit

**Comment by talregev** [comment]
Simplify version update:

* [x]  Add GUI element in Advanced settings to display current content of gui_rpc_auth.cfg file
* [x]  If gui_rpc_auth.cfg file doesn't exist or empty - generate new password (32 characters long)
* [x]  By default, this password should be hidden (with dots like password)
* [x]  When user clicks on this element, dialog box should be displayed with current password unhidden
* [x]  In this dialog user should be able to set new password, this password should be written to gui_rpc_auth.cfg file
* [ ]  Empty password is not allowed, user should be notified about this
* [ ]  ~~If user sets new password while client is already running with remote monitoring on, user should be informed that after new password set, client will be restarted and all running tasks will be interrupted~~
* [ ]  ~~If user confirms change password when remote monitoring is on, restart client~~
* [ ]  ~~NICE TO HAVE: Add protection to both buttons ('Show' password and 'Edit' password) to confirm act with default device protection mechanism: password or Fingerprint/FaceID if available).~~

The strikethrough it for next task, because now I don't have a state of remote boinc. Only change this file.
I have problem with empty password.
When the user will enter empty password, I prevent it from update, but it inconsistent with the GUI.
The GUI show empty password, while the file have the previous password. And user will not see until he exit the settings and enter them again.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 1]
```diff
@@ -32,10 +33,17 @@ import edu.berkeley.boinc.utils.setAppTheme
...
+import java.io.File
+import java.util.concurrent.ThreadLocalRandom
+import kotlin.streams.asSequence
 
 class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {
     private val hostInfo = BOINCActivity.monitor!!.hostInfo // Get the hostinfo from client via RPC
     private val prefs = BOINCActivity.monitor!!.prefs
+    private val charPool : List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')
+    private val passwordLength = 32
+    private var autKey = ""
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:45]: `autKey` -> `authKey`

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 2]
```diff
@@ -56,6 +64,15 @@ class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedP
             sharedPreferences.edit { putString("deviceName", hostInfo.domainName) }
         }
 
+        if ("authenticationKey" !in sharedPreferences) {
+            autKey = readAutFileContent()
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:68]: `readAutFileContent` -> `readAuthFileContent`

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 3]
```diff
@@ -32,10 +33,17 @@ import edu.berkeley.boinc.utils.setAppTheme
...
 import kotlinx.coroutines.launch
+import java.io.File
+import java.util.concurrent.ThreadLocalRandom
+import kotlin.streams.asSequence
 
 class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {
     private val hostInfo = BOINCActivity.monitor!!.hostInfo // Get the hostinfo from client via RPC
     private val prefs = BOINCActivity.monitor!!.prefs
+    private val charPool : List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')
+    private val passwordLength = 32
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:44]: Should be constant

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 4]
```diff
@@ -32,10 +33,17 @@ import edu.berkeley.boinc.utils.setAppTheme
...
 import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.launch
+import java.io.File
+import java.util.concurrent.ThreadLocalRandom
+import kotlin.streams.asSequence
 
 class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {
     private val hostInfo = BOINCActivity.monitor!!.hostInfo // Get the hostinfo from client via RPC
     private val prefs = BOINCActivity.monitor!!.prefs
+    private val charPool : List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:43]: Could be constant

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 5]
```diff
@@ -56,6 +64,15 @@ class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedP
             sharedPreferences.edit { putString("deviceName", hostInfo.domainName) }
         }
 
+        if ("authenticationKey" !in sharedPreferences) {
+            autKey = readAutFileContent()
+            if (autKey.isEmpty()) {
+                autKey = generateRandomString(passwordLength)
+                writeAutFileContent(autKey)
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:71]: `writeAutFileContent` -> `writeAuthFileContent`

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 6]
```diff
@@ -191,6 +211,18 @@ class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedP
                 lifecycleScope.launch { writeClientPrefs(prefs) }
             }
 
+            "authenticationKey" -> {
+                val currentAutKey = sharedPreferences.getString(key, "")!!
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:215]: `currentAutKey` -> `currentAuthKey`

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt [hunk 7]
```diff
@@ -246,4 +278,26 @@ class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedP
...
+                .joinToString("")
+    }
+
+    // Return the password in asterisks
+    private fun setAsterisks(length: Int): String {
+        return "*".repeat(length)
+    }
+
+    private fun readAutFileContent(): String {
+        val autFile = File(BOINCActivity.monitor!!.authFilePath)
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:296]: `autFile` -> `authFile`

**Review by AenBleidd — CHANGES_REQUESTED** [review]

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt (hunk 4 above)
Comment (talregev) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:43]: In kotlin constant it mean static member to the class, then I need to do companion object. 
It more code and we don't gain performance out of that.

Comment (talregev) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt:43]: val also mean that you cannot change the value. so it already constant.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt (hunk 1 above)
Comment (talregev) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt]: refactor

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt (hunk 6 above)
Comment (talregev) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt]: refactor

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt (hunk 7 above)
Comment (talregev) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt]: refactor

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt (hunk 2 above)
Comment (talregev) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SettingsFragment.kt]: refactor

**Review by AenBleidd — APPROVED** [review]

**Pull request merged**